x/fibre/spec.md

```diff
@@ -0,0 +1,221 @@
+
+## State
+
+The `x/fibre` module stores the following data:
+
+### FibreProviderInfo
+
+```protobuf
+message FibreProviderInfo {
+  // ip_address is the IP address where users can access the fibre service
+  string ip_address = 1;
+}
+```
+
+### Store Keys
+
+- `0x01 | ValAddress -> ProtocolBuffer(FibreProviderInfo)`: Maps validator address to fibre provider info
+
+## Messages
+
+### MsgSetFibreProviderInfo
+
+Allows a validator to set or update their fibre provider information.
+
+```protobuf
+message MsgSetFibreProviderInfo {
+  // validator_address is the operator address of the validator
+  string validator_address = 1;
+  // ip_address is the IP address for the fibre service (max 45 characters for IPv6)
+  string ip_address = 2;
+}
+```
+
+**Validation Rules:**
+- `validator_address` must be a valid validator operator address
+- `ip_address` must be non-empty and ≤ 45 characters
+- Signer must match the validator address
+- Validator must be in the active set
+
+### MsgRemoveFibreProviderInfo
+
+Allows removal of fibre provider information for validators not in the active set.
+
+```protobuf
+message MsgRemoveFibreProviderInfo {
+  // validator_address is the operator address of the validator to remove
+  string validator_address = 1;
+  // remover_address is the address of the account requesting removal
+  string remover_address = 2;
+}
+```
+
+**Validation Rules:**
+- `validator_address` must be a valid validator operator address
+- `remover_address` must be a valid account address
+- Validator must NOT be in the active set
+- Provider info must exist for the validator
+
+## Events
+
+### EventSetFibreProviderInfo
+
+Emitted when a validator sets or updates their fibre provider information.
+
+```protobuf
+message EventSetFibreProviderInfo {
+  // validator_address is the operator address of the validator
+  string validator_address = 1;
+  // ip_address is the IP address for the fibre service
+  string ip_address = 2;
+}
+```
+
+### EventRemoveFibreProviderInfo
+
+Emitted when fibre provider information is removed.
+
+```protobuf
+message EventRemoveFibreProviderInfo {
+  // validator_address is the operator address of the validator
+  string validator_address = 1;
+  // remover_address is the address of the account that requested removal
+  string remover_address = 2;
+}
+```
+
+## Queries
+
+### QueryFibreProviderInfo
+
+Query fibre provider information for a specific validator.
+
+**Request:**
+```protobuf
+message QueryFibreProviderInfoRequest {
+  // validator_address is the operator address of the validator
+  string validator_address = 1;
+}
+```
+
+**Response:**
+```protobuf
+message QueryFibreProviderInfoResponse {
+  // info contains the fibre provider information
+  FibreProviderInfo info = 1;
+  // found indicates if the validator has registered info
+  bool found = 2;
+}
+```
+
+### QueryAllActiveFibreProviders
+
+Query fibre provider information for all validators in the active set.
```

Comment:
Ideally there should be another message to query diff from previous requested state of ValSet. Similar git fetch. Request would need to point to previous state hash / block heigh and server will be able to server just diff if there is any. This would save quite some bandwidth on periodic updates of ValSet on client (for both client and server)

Reply:
> Ideally there should be another message to query diff from previous requested state of ValSet.

doing this is actually a very large change, as we'd need to enforce old copies are kept in state or that all rpc nodes have past copies of the state.

naively, optimizing for bandwidth atm seems like a premature optimization since these nodes will be uploading and downloading large amount of data to and from validators. imo, it seems like the diff can be managed effectively by re-querying for specific unknown validator's FibreProviderInfo as they join the set since the light client is already tracking the validator set to verify the header, or simply by re-querying this endpoint to get the latest set.